File: apps/web/src/api/faculty.ts
import { apiFetch } from "../lib/apiClient";

export interface FacultyProfile {
  id: string;
  user_id: string;
  employee_id: string;
  department_id?: string;
  designation: string;
  workload_hours: number;
  qualification: string;
  join_date?: string;
  status: string;
  created_at: string;
}

export interface LeaveRequest {
  id: string;
  faculty_id: string;
  start_date: string;
  end_date: string;
  reason: string;
  status: string; // Pending | Approved | Rejected
  created_at: string;
}

export async function apiGetFaculty(skip = 0, limit = 50): Promise<FacultyProfile[]> {
  return apiFetch<FacultyProfile[]>(`/api/v1/faculty?skip=${skip}&limit=${limit}`);
}

export async function apiGetFacultyMember(id: string): Promise<FacultyProfile> {
  return apiFetch<FacultyProfile>(`/api/v1/faculty/${id}`);
}

export async function apiCreateFaculty(data: Partial<FacultyProfile>): Promise<FacultyProfile> {
  return apiFetch<FacultyProfile>("/api/v1/faculty", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(data),
  });
}

export async function apiUpdateFaculty(id: string, data: Partial<FacultyProfile>): Promise<FacultyProfile> {
  return apiFetch<FacultyProfile>(`/api/v1/faculty/${id}`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(data),
  });
}

export async function apiGetFacultyLeaves(facultyId: string): Promise<LeaveRequest[]> {
  return apiFetch<LeaveRequest[]>(`/api/v1/faculty/${facultyId}/leaves`);
}

export async function apiApplyLeave(facultyId: string, data: Partial<LeaveRequest>): Promise<LeaveRequest> {
  return apiFetch<LeaveRequest>(`/api/v1/faculty/${facultyId}/leaves`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ ...data, faculty_id: facultyId }),
  });
}

export async function apiApproveLeave(leaveId: string, approve: boolean): Promise<LeaveRequest> {
  return apiFetch<LeaveRequest>(`/api/v1/faculty/leaves/${leaveId}/status`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ status: approve ? "Approved" : "Rejected" }),
  });
}

File: backend/app/schemas/faculty.py
from pydantic import BaseModel, Field, ConfigDict
from datetime import datetime, date
from typing import Optional

class FacultyBase(BaseModel):
    user_id: str
    employee_id: str = Field(..., min_length=2, max_length=30)
    department_id: Optional[str] = None
    designation: str = Field(..., min_length=2, max_length=100)
    workload_hours: int = Field(14, ge=0, le=40)
    qualification: str = Field("B.E.", min_length=2, max_length=100)
    join_date: Optional[date] = None
    status: str = Field("Active", description="Active | On Leave | Terminated")

class FacultyCreate(FacultyBase):
    pass

class FacultyUpdate(BaseModel):
    designation: Optional[str] = None
    workload_hours: Optional[int] = None
    qualification: Optional[str] = None
    status: Optional[str] = None

class FacultyResponse(FacultyBase):
    id: str
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)

class LeaveBase(BaseModel):
    faculty_id: str
    start_date: date
    end_date: date
    reason: str = Field(..., min_length=5, max_length=255)

class LeaveCreate(LeaveBase):
    pass

class LeaveStatusUpdate(BaseModel):
    status: str = Field(..., description="Pending | Approved | Rejected")

class LeaveResponse(LeaveBase):
    id: str
    status: str
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)
